local dltime = dlimport.load("@time/time")
print(`path for {nameof(dltime)} is {dltime.path}`)
local initmodule: ()->() = dltime:importcfunction("initmodule")
print(`{nameof(initmodule)} is {typeof(initmodule)}`)
assert(typeof(initmodule) == 'function')
local module = require('dummy_module')
assert(module.name == 'init')
local submodule = require('dummy_module/submodule')
assert(submodule.name == 'submodule')
local cc = require('console_codes')
print(`{cc.brightgreen}{script.name} tests successful{cc.reset}`)
