local this_thread = coroutine.running()
print(nameof(this_thread), this_thread, coroutine.status(this_thread))
local co = coroutine.create(function(word: string)
    print(`im a coroutine {word}`)
end)
print(`precomp: {nameof(this_thread)}`)
task.delay_until(task.this, function()
    print("Script is finished")
end)
task.spawn(function()
    task.wait_until(this_thread)
end)

local d = task.delay(5, co, "for real")
task.spawn(function()
    task.wait_until(d)
    print('i waited')
end)
local s = task.spawn(function(str)
    task.delay_until(task.this, function()
        print("after spawn")
    end)
    print(`status {nameof(this_thread)} = {coroutine.status(this_thread)}`)
    print(`yo {str}`)
    task.defer(function()
        print('frfr')
    end)
end, "world")
print(`spawn {s}`)
collectgarbage('collect')
local e = task.defer(function(str)
    print(`bye bye {str}`)
end, "you")

local v = task.defer(function(str)
    task.wait(3)
    print(`bye {str}`)
end, "world")
this_thread = coroutine.running()
print(nameof(this_thread), this_thread, coroutine.status(this_thread))
print(`defer {v}`)
task.wait(6)
local cc = require('console_codes')
print(`{cc.brightgreen}{script.name} tests successful{cc.reset}`)
