local time = require('@require/qol2/time')
print.text_color = 'blue'
local co = coroutine.create(function(word: string)
    print(`im a coroutine {word}`)
end)

task.delay(1, co, "for real")
local s = task.spawn(function(str)
    print(`yo {str}`)
    task.defer(function()
        print('frfr')
    end)
end, "world")
print(`spawn {s}`)
local e = task.defer(function(str)
    print(`bye bye {str}`)
end, "you")

local v = task.defer(function(str)
    print(`bye {str}`)
end, "world")
print(`defer {v}`)
