local time = require('@require/qol2/time')
print.text_color = 'blue'
local this_thread = coroutine.running()
print(nameof(this_thread), this_thread, coroutine.status(this_thread))
local co = coroutine.create(function(word: string)
    print(`im a coroutine {word}`)
end)
task.after_this(function()
    print("Script is finished")
end)

local d = task.delay(5, co, "for real")
task.after(d, function()
    print(coroutine.status(d))
    print('i waited')
end)
local s = task.spawn(function(str)
    task.after_this(function()
        print("after spawn")
    end)
    print(`status {nameof(this_thread)} = {coroutine.status(this_thread)}`)
    print(`yo {str}`)
    task.defer(function()
        print('frfr')
    end)
end, "world")
print(`spawn {s}`)
collectgarbage('collect')
local e = task.defer(function(str)
    print(`bye bye {str}`)
end, "you")

local v = task.defer(function(str)
    print(`bye {str}`)
end, "world")
this_thread = coroutine.running()
print(nameof(this_thread), this_thread, coroutine.status(this_thread))
task.wait(7)
print(`defer {v}`)
